doc/qos/reclaim-reserved-buffer.md

```diff
@@ -0,0 +1,696 @@
+
+## 5 Requirements ##
+
+The requirement is to reclaim the reserved buffer for admin down ports, including:
+
+- Buffer reserved SONiC configuration
+  - BUFFER_PG
+  - BUFFER_QUEUE
+  - BUFFER_PORT_INGRESS_PROFILE_LIST / BUFFER_PORT_EGRESS_PROFILE_LIST
+- Buffer reserved by SDK/SAI for objects that are not configured by SONiC, out of scope of this design but will be addressed by SAI
+  - management PG when the port is shutdown
+  - If the SONiC configures buffer pools
+    - The default lossy priority group configured by SDK
+
+## 6 Architecture Design ##
+
+### 6.1 The way to set reserved buffer to zero ###
+
+Currently, the reserved size of a buffer object is set to zero when it is removed from `BUFFER_PG` or `BUFFER_QUEUE` table. A way to reclaim buffer is
+
+- SONiC to remove objects of admin-down ports from `BUFFER_PG` and `BUFFER_QUEUE` tables
+- SAI to set reserved size of buffer objects to zero on removing them.
+
+However, this creates inconsistency. Consider the following scenarios:
+
+1. System starting flow. SAI will not touch buffer objects if there is no buffer related configuration applied from SONiC, leaving them as the SDK default value. As a result, for any buffer object,
+   - There is no buffer configuration in SONiC
+   - The reserved buffer size in the ASIC is the SDK default value, some of which are not zero
+2. System started, an existing buffer object is removed. SONiC notifies SAI by setting profile to `SAI_NULL_OBJECT_ID`. SAI will set the reserved size of corresponding buffer object to zero. As a result, for the buffer object,
+   - There is no buffer configuration in SONiC
+   - The reserved buffer size in the ASIC is zero
+
+For some of the buffer objects, the SDK default reserved size is not zero. This is to make sure the system works correctly without any buffer configuration in SONiC.
+Now we have same SONiC configuration in 1 and 2 but different reserved buffer size in the ASIC.
+
+To make it clear and consistent, we need the following solution:
+
+- Introduce a series of buffer profiles whose `size` is zero and `dynamic_th` is -8. This means it will neither reserve buffer nor occupy shared buffer pool. This kind of profiles will be called as `zero buffer profile` in the following section.
+- The buffer objects will be configured to:
+  - SDK default value, if there is no buffer profile configured for the object in SONiC
+  - SDK default value, if the buffer object is removed from SONiC
+  - Zero, only if the object is configured with a zero profile
+
+To achieve it the following steps will be taken:
+
+1. Zero buffer profiles should be defined for ingress/egress and lossless/lossy traffic.
+2. Currently, there is no buffer object configured on admin-down ports. The zero profiles should be configured explicitly on admin-down ports.
+3. `ingress_lossy_profile` is for the ingress lossy priority group.
```

Comment:
Need to find a way to indicate whether a platform supports `empty profile`. Candidates include
- if the `empty_profile` is provided by the vendor, it is supported
  - pros: easy to implement. no timing hazard
  - cons: the `empty_profile` needs to be loaded even if it is not used.
- query SAI capability.
  - cons: 
    - SAI attribute can only be fetched by orchagent and then exposed to state db. 
    - no guarantee that the capability is available when buffer manager starts. need to handle both conditions.